fix(config): prohibit multiline interpolation [WIP]

pront opened the pull request:
<!--
  Your PR title must conform to the conventional commit spec:
  https://www.conventionalcommits.org/en/v1.0.0/

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs, revert
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/vectordotdev/vector/blob/master/.github/workflows/semantic.yml#L31
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

## Summary
<!-- Please provide a brief summary about what this PR does.
This should help the reviewers give feedback faster and with higher quality. -->

> [!WARNING]
> This PR has a very wide surface area. It would be best to split it into individual PRs.
> Also, it is not actively worked on. 
> YAML config loading seems to be working fine, but TOML configs have a few bugs. 

The envinronment variable (and the secrets) resolution now happens after the config string is parsed into a TOML table. The interpolation now happens on the table keys and values. As a side effect, this fixes a bug where comments referring to env vars (or secrets) that don't exist caused a config build error.

## Change Type
- [x] Bug fix
- [ ] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [x] Yes
- [ ] No

Injecting whole blocks now results in error e.g. 

```shell
export SOURCES_BLOCK="sources:\"
  demo:
    type: demo_logs
    format: json
    interval: 1
```

```yaml
${SOURCES_BLOCK}
```

The config above will fail to load.

## How did you test this PR?
<!-- Please describe your testing plan here.
Sharing information about your setup and the Vector configuration(s) you used (when applicable) is highly recommended.
Providing this information upfront will facilitate a smoother review process. -->

Vector config:

```yaml
secret:
  backend_1:
    type: file
    path: /Users/pavlos.rontidis/CLionProjects/vector/pront/data/top_secret.json

# ${IN_COMMENT_BUT_DOES_NOT_EXIST}

sources:
  demo_logs_1:
    type: demo_logs
    format: json
    interval: 60
    # noop: "SECRET[i_dont_exist]"

transforms:
  t0:
    inputs:
      - demo_logs_1
    type: "remap"
    source: |
      .host = "${HOSTNAME}"
      .environment = "${ENV:?emv must be supplied}"
      .tenant = "${TENANT:-undefined}"
      .day = "SECRET[backend_1.day]"


sinks:
  s0:
    type: "SECRET[backend_1.type]"
    inputs: [ "t0" ]
    encoding:
      codec: json
      json:
        pretty: true
```

Output:
```text
{
  "day": "monday",
  "environment": "${ENV:?emv must be supplied}",
  "host": "${HOSTNAME}",
  "message": "{\"host\":...",
  "service": "vector",
  "source_type": "demo_logs",
  "tenant": "${TENANT:-undefined}",
  "timestamp": "2025-05-05T20:52:42.024064Z"
}
```

## Does this PR include user facing changes?

- [ ] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Notes
- Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- Do not hesitate to use `@vectordotdev/vector` to reach out to us regarding this PR.
- The CI checks run only after we manually approve them.
  - We recommend adding a `pre-push` hook, please see [this template](https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#Pre-push).
  - Alternatively, we recommend running the following locally before pushing to the remote branch:
    - `cargo fmt --all`
    - `cargo clippy --workspace --all-targets -- -D warnings`
    - `cargo nextest run --workspace` (alternatively, you can run `cargo test --all`)
    - `./scripts/check_changelog_fragments.sh`
- After a review is requested, please avoid force pushes to help us review incrementally.
  - Feel free to push as many commits as you want. They will be squashed into one before merging.
  - For example, you can run `git merge origin master` and `git push`.
- If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `cargo vdev build licenses` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

## References

<!-- Please list any issues closed by this PR. -->


- closes: https://github.com/vectordotdev/vector/issues/797
- closes: https://github.com/vectordotdev/vector/issues/22145
- Ref: https://github.com/vectordotdev/vector/issues/17343 (need to double check if this is resolved by this PR)

pront committed:
fix(config): disallow env var block injection and ignore comments before interpolation"

pront committed:
add test for block in invalid ctx

pront commented on src/config/loading/interpolation.rs:
```diff
@@ -0,0 +1,297 @@
+use std::{collections::HashMap, sync::LazyLock};
```
Moved `vars.rs` to this location. But kept the same use/export.

pront committed:
improve changelog

pront commented on src/config/loading/loader.rs:
```diff
@@ -297,3 +295,31 @@ pub(super) fn deserialize_table<T: serde::de::DeserializeOwned>(
+    deserialize_table(table)
+}
+
+pub fn resolve_environment_variables(table: Table) -> Result<Table, Vec<String>> {
```
This was extracted from `prepare_input`

pront committed:
fix bug after bad refactoring, cleanup

jszwedko commented on changelog.d/multiline_interpolation_change.breaking.md:
````diff
@@ -0,0 +1,20 @@
+${SOURCES_BLOCK}
+```
+
+The config above will fail to load.
````
I'd suggest we include mention of the workaround users can use here.

jszwedko commented on changelog.d/multiline_interpolation_change.breaking.md:
```diff
@@ -0,0 +1,20 @@
+The env var (and secrets) resolution now happens after the config string is parsed into a TOML table.
```
Is the config always parsed into a TOML table? This explains the weird behavior I've seen before where I get TOML errors for YAML configs. 😅

pront replied:
Right, internally we call `format::deserialize` which supports all formats and then we always convert to a TOML table after this step. I would expect this to translate to a config builder (out of scope for this PR).

jszwedko commented on changelog.d/multiline_interpolation_change.breaking.md:
```diff
@@ -0,0 +1,20 @@
+The env var (and secrets) resolution now happens after the config string is parsed into a TOML table.
```
```suggestion
The env var and secrets resolution now happens after the config string is parsed into a TOML table.
```

I think both are equally important to note.

jszwedko commented on changelog.d/multiline_interpolation_change.breaking.md:
```diff
@@ -0,0 +1,20 @@
+The env var (and secrets) resolution now happens after the config string is parsed into a TOML table.
+As a side effect, this fixes a bug where comment lines referring to env vars (or secrets) that don't exist caused a config build error.
```
```suggestion
As a side effect, this fixes a bug where comment lines referring to env vars or secrets that don't exist caused a config build error.
```

jszwedko commented on src/config/loading/loader.rs:
```diff
@@ -297,3 +298,31 @@ pub(super) fn deserialize_table<T: serde::de::DeserializeOwned>(
+
+pub fn resolve_environment_variables(table: Table) -> Result<Table, Vec<String>> {
+    let mut vars = std::env::vars().collect::<HashMap<_, _>>();
+    if !vars.contains_key("HOSTNAME") {
```
TIL this special case was here.

jszwedko reviewed (commented):
Nice! I'm glad to see us finally addressing this rough edge.

I think this works for string fields, but it doesn't seem to interpolate non-string fields correctly. For example:

```yaml
sources:
  source0:
    type: demo_logs
    count: ${COUNT}
    format: json
    decoding:
      codec: json
    framing:
      method: bytes
    interval: 1.0
sinks:
  sink0:
    inputs:
    - source0
    type: console
    encoding:
      codec: json
```

With `COUNT=1` gives:

```
2025-05-06T21:40:58.830829Z  INFO vector::app: Log level is enabled. level="info"
2025-05-06T21:40:58.831986Z  INFO vector::app: Loading configs. paths=["/tmp/tmp.yaml"]
2025-05-06T21:40:58.838663Z ERROR vector::cli: Configuration error. error=invalid type: string "1", expected usize



in `sources.source0`
```

pront committed:
handle additional types (i64, f64, bool)

pront committed:
add workaround to changelog

pront committed:
add test for multiple secret types

pront committed:
chore: update generated license files

pront commented on src/config/loading/secret.rs:
```diff
@@ -88,11 +89,10 @@ impl SecretBackendLoader {
 }
 
```
Note: Currently (on master) we don't support multiple types for secrets.

```
2025-05-07T14:38:00.776983Z ERROR vector::cli: Configuration error. error=Error while retrieving secret from backend "backend_1": invalid type: integer `1`, expected a string at line 1 column 10.
```

However, `interpolate_secrets` supports multiple types. I consider this out of scope for this PR but we could go further and allow e.g. integer secrets.

pront commented:
> With `COUNT=1` gives:
> 
> ```
> 2025-05-06T21:40:58.830829Z  INFO vector::app: Log level is enabled. level="info"
> 2025-05-06T21:40:58.831986Z  INFO vector::app: Loading configs. paths=["/tmp/tmp.yaml"]
> 2025-05-06T21:40:58.838663Z ERROR vector::cli: Configuration error. error=invalid type: string "1", expected usize


Fixed and added tests to cover this.

pront committed:
markdown lint

jszwedko reviewed (commented):
Nice, thanks. This fixes my first example. I think the optimistic parsing as integers, though, isn't quite right. It doesn't seem to handle quoted strings where the value is parsable as an integer. For example, this doesn't work:

```
sources:
  source0:
    type: demo_logs
    count: ${COUNT}
    format: shuffle
    lines: ["${LINE}"]
    framing:
      method: bytes
    interval: 1.0
sinks:
  sink0:
    inputs:
    - source0
    type: console
    encoding:
      codec: json
```

It gives:

```
COUNT=1 LINE=1234 ./target/debug/vector -c /tmp/tmp.yaml
2025-05-07T15:31:36.925469Z  INFO vector::app: Log level is enabled. level="info"
2025-05-07T15:31:36.926534Z  INFO vector::app: Loading configs. paths=["/tmp/tmp.yaml"]
2025-05-07T15:31:36.932941Z ERROR vector::cli: Configuration error. error=invalid type: integer `1234`, expected a string

in `sources.source0`
```

pront committed:
coerce

pront committed:
chore(dev): vdev fmt

pront committed:
chore(dev): vdev fmt

pront committed:
chore(dev): cargo vdev build licenses

pront commented:
So there are some tests failures:
1. existing e2e regression
2. this PR broke some tests (no_proxy is set to a string but schema specifies array of strings)

Will fix (1) in a separate PR
Will fix (2) in this PR (after 1 is fixed)

pront committed:
permissive conversion to array, for cases like proxy.no_config='x' (converted to array eventually)

pront committed:
add guard to skip secret processing

pront committed:
push the coercion logic deeper in the stack

pront committed:
remove dbg statement

pront commented:
Still testing this. I found at least one bug for TOML configs. YAML configs seems to be working fine.

pront committed:
wip

pront committed:
better coercion test

pront committed:
fix any_of and one_of hnadling and separate them

pront committed:
dbg commit

pront committed:
move resoltion in Process

pront committed:
Debug
